### ISSUE TITLE
Issue tracker for Full GNOME 46 support 

### ISSUE THREAD

**Issue by Totto16** [oae/gnome-shell-pano#271]
Since #264 adds support for gnome 46 it should be good in terms of GJS changes, but while working on #265 I noticed some changes in the gnome-shell UI libraries, we also use. There are some changes that I, and presumably others didn't notice, since e.g. `Source` from [here](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/js/ui/messageTray.js) doesn't have `showNotification` anymore since GNOME 46, but it's called `addNotification`. But the types suggested it before, so while fixing types over at https://github.com/gjsify/gnome-shell we should check, if we need to change more types to check for gnome 45 / 46 compatibility at runtime. 

Task list:
- [ ] fix types for gnome 46 and 45 compatibility #264
- [ ] remove highlights.js code highlighting, and make it something users have to install manually or user an alternative (e.g. `source-highlight` (see https://github.com/oae/gnome-shell-pano/pull/264#issuecomment-2026022132), this is to comply with EGO guidlines

Optional:
- [ ] #266 - fix long standing bug

**Comment by RushingAlien** [comment]
I would like to add that due to direct scanout for full screen windows under Wayland, pano will fail do display when a window is fullscreen, the solution for this is using `Meta.disable_unredirect_for_display()` when the pano dialog/panel pops up. References on how to do this can be found in https://github.com/G-dH/window-thumbnails/commit/bb875073c8584bb1fbac3eb0ebf123384d4abdd2 and  https://github.com/icedman/search-light/commit/25b43eab35adcaf50f5e3c58a84d9af8a29745cd. 

This is only needed until GNOME supports overlay planes which will allow layered and non-fullscreen direct scanout.


Note from window-thumbnail dev about disable_unredirect: https://github.com/G-dH/window-thumbnails/issues/7#issuecomment-1997187140

**Comment by Totto16** [comment]
> I would like to add that due to direct scanout for full screen windows under Wayland, pano will fail do display when a window is fullscreen, the solution for this is using `Meta.disable_unredirect_for_display()` when the pano dialog/panel pops up. References on how to do this can be found in [G-dH/window-thumbnails@bb87507](https://github.com/G-dH/window-thumbnails/commit/bb875073c8584bb1fbac3eb0ebf123384d4abdd2) and [icedman/search-light@25b43ea](https://github.com/icedman/search-light/commit/25b43eab35adcaf50f5e3c58a84d9af8a29745cd).
> 
> This is only needed until GNOME supports overlay planes which will allow layered and non-fullscreen direct scanout.
> 
> Note from window-thumbnail dev about disable_unredirect: [G-dH/window-thumbnails#7 (comment)](https://github.com/G-dH/window-thumbnails/issues/7#issuecomment-1997187140)

Thanks for the suggestion 👍🏼  I will test that and also fix it 👍🏼

**Comment by marcinjahn** [comment]
@Totto16 The README mentions Gnome 46 as being supported, while EGO is still on Gnome 45. I assume that I could just get the latest zip from Releases and it will work on G46, right? It seems that the only "issue" there is highlight.js still being embedded.

**Comment by Totto16** [comment]
@marcinjahn Correct, the "issue" with highlight.js is, that EGO doesn't want to have that too much third party dependencies in published extensions. So The master branch fully supports GNOME 46, but for a release on EGO I have to complete #278 (which might take some time).

To install it locally, follow the steps [here](https://github.com/oae/gnome-shell-pano?tab=readme-ov-file#development) 

You have to download the latest files from the master branch tough, the latest release doesn't support gnome 46 yet and we didn't make a new release yet 😓

**Comment by marcinjahn** [comment]
Sure, thanks for the clarification!

**Comment by theGeekyLad** [comment]
Does this issue also pertain to GNOME 46.1 (currently breaking)?

**Comment by Totto16** [comment]
> Does this issue also pertain to GNOME 46.1 (currently breaking)?

Not really, this is mainly for Gnome 46.0 support (and the EGO release for that) . 

That said, just open a new issue with the exact details and I'll look into it separately. The EGO topic might take a while. 
But fixing compatibility for gnome 46.1 would take a higher priority. 

This is really interesting, since minor releases shouldn't break anything, but I'm not surprised, gnome 45.3 broke some things too 🤦🏼‍♂️

**Comment by oae** [comment]
I tested the master branch on 46.1 and saw no issues.

**Comment by Totto16** [comment]
> I tested the master branch on 46.1 and saw no issues.

Interesting, then he probably meant the public ego release 😓 
Or am I wrong @theGeekyLad

**Comment by talesam** [comment]
> Testei a filial principal em 46.1 e não vi problemas.

I am one of the Biglinux Devs and I take care of the Gnome version, it is on 46.1 and is based on Manjaro and it is not working, it says that the extension is not compatible with version 46.1

**Comment by Totto16** [comment]
> > Testei a filial principal em 46.1 e não vi problemas.
> 
> I am one of the Biglinux Devs and I take care of the Gnome version, it is on 46.1 and is based on Manjaro and it is not working, it says that the extension is not compatible with version 46.1

The version on EGO doesn't yet support 46 due to some concern on some third party libraries from the side of the EGO reviewers, we are working on that, see #278

but you can self "compile" (transpile) the master branch and than it works

**Comment by Malix-off** [comment]
Is there an ETA as of when pano will be available for Gnome 46 ?

**Comment by Totto16** [comment]
> Is there an ETA as of when pano will be available for Gnome 46 ?

I have started to do the needed work, quite a while ago, (see #278 ) but I had some "more important" things to do, so I didn't have time for quite a while, I can use the extension by self building it so it also climbed down my priority list 😓  (which is kinda egostic and not a good excuse 😓 ).

I see that there is "high" demand and that Gnome 46 is out on many distros for a while (almost a month on Ubuntu, Fedora etc.) 
So I am trying my best to schedule the needed work for as soon as possible 😓

**Comment by logicito** [comment]
> > Is there an ETA as of when pano will be available for Gnome 46 ?
> 
> I have started to do the needed work, quite a while ago, (see #278 ) but I had some "more important" things to do, so I didn't have time for quite a while, I can use the extension by self building it so it also climbed down my priority list 😓 (which is kinda egostic and not a good excuse 😓 ).
> 
> I see that there is "high" demand and that Gnome 46 is out on many distros for a while (almost a month on Ubuntu, Fedora etc.) So I am trying my best to schedule the needed work for as soon as possible 😓

Thank you very much, there is no better clipboard manager than Pano

**Comment by daethlor** [comment]
> @marcinjahn Correct, the "issue" with highlight.js is, that EGO doesn't want to have that too much third party dependencies in published extensions. So The master branch fully supports GNOME 46, but for a release on EGO I have to complete #278 (which might take some time).
> 
> To install it locally, follow the steps [here](https://github.com/oae/gnome-shell-pano?tab=readme-ov-file#development)
> 
> You have to download the latest files from the master branch tough, the latest release doesn't support gnome 46 yet and we didn't make a new release yet 😓

On 24.04, when trying to self build I get:
```
00h00m00s 0/0: : ERROR: [Errno 2] No such file or directory: 'install'
00h00m00s 0/0: : ERROR: [Errno 2] No such file or directory: 'build'
```
Would love to have even partial usability on 24.04 at this point.

**Comment by back-lacking** [comment]
is there an eta on gnome 45/46 support or is it currently still unknown?

**Comment by Totto16** [comment]
I published a pre-release, that has to be installed manually, but at least you don't need to self-built the extension. 
Until I finally fix everything, and we can release it on EGO, it might take a while😓

**Comment by 7system7** [comment]
OMG. This extension is (sadly) irreplaceable. Thank you.

**Comment by johnstonesnow** [comment]
> I published a pre-release, that has to be installed manually, but at least you don't need to self-built the extension. Until I finally fix everything, and we can release it on EGO, it might take a while😓

Hi, is there any chance someone could explain how to install it please? I am on Fedora 40 and lost without Pano. thanks

**Comment by Vinnl** [comment]
@johnstonesnow Information about the prelease can be found [here](https://github.com/oae/gnome-shell-pano/releases/tag/v23-alpha).

From that page, download `pano@elhan.io.zip`. Open Files, click the path bar at the top and enter `~/.local/share/gnome-shell/extensions/` and press <kbd>Enter</kbd>. Copy the downloaded file into that folder, then open it to extract its contents into that folder. Next time you login, the extension should work.

**Comment by johnstonesnow** [comment]
Ah, great thanks. 
Forgive my ignorance, but if I uninstall the old one (not even sure how but will look that up), wont I need to install the new one? I was under the impression this was just a patch to fix a current installation,maybe its a full manual installation?
thanks again

**Comment by Vinnl** [comment]
It is indeed a full manual installation - extracting the `.zip` into that directory _is_ installing it (the prerelease).

**Comment by johnstonesnow** [comment]
Great, will give it a whirl. thanks

**Comment by johnstonesnow** [comment]
PS Just had a thought. I am thinking of trying out KDE Plasma DE on Fedora. I have never used it, fairly new to Linux overall. Am I right in thinking, if i were to switch to KDE, Pano will probably work (as it seems the issue is with Gnome DE)?
thanks

**Comment by logicito** [comment]
> PS Just had a thought. I am thinking of trying out KDE Plasma DE on Fedora. I have never used it, fairly new to Linux overall. Am I right in thinking, if i were to switch to KDE, Pano will probably work (as it seems the issue is with Gnome DE)? thanks

Pano is a Gnome extension, it only works with Gnome

**Comment by Totto16** [comment]
> PS Just had a thought. I am thinking of trying out KDE Plasma DE on Fedora. I have never used it, fairly new to Linux overall. Am I right in thinking, if i were to switch to KDE, Pano will probably work (as it seems the issue is with Gnome DE)? thanks

No, Pano only works with Gnome. KDE is another DE (Desktop Environment). 
Pano is a gnome extension and therefore only works with Gnome. It also uses a lot of Gnome specific features and on KDE there are other clipboard managers that you can use 😓

**Comment by johnstonesnow** [comment]
Oops, this is the part where I look dumb :D 
Thanks. I just played with KDE a bit and whilst it has some VERY nice features, it has others I don't like as much as Gnome's equivalent. I am gonna stick with Gnome, partly just to keep Pano :)

**Comment by thyttan** [comment]
### BUG:
- Single-finger tap (left mouse button click) on icon opens clipboard AND top bar menu.
- Tap again hides the top bar menu.
- Tap a third time hides the clipboard.

### EXPECTED: 
To only display the clipboard. (One click should show. If I click again it should hide)

### SYSTEM
Fedora Linux 40

Can anyone confirm they see this too?

**Comment by Totto16** [comment]
@lorenzo0932  @Dkoeddnsnp  Could you file a separate issue for the bug. Then I'll look into it ❤️

**Comment by Totto16** [comment]
> ### BUG:
> 
>     * Single-finger tap (left mouse button click) on icon opens clipboard AND top bar menu.
> 
>     * Tap again hides the top bar menu.
> 
>     * Tap a third time hides the clipboard.
> 
> 
> ### EXPECTED:
> 
> To only display the clipboard. (One click should show. If I click again it should hide)
> ### SYSTEM
> 
> Fedora Linux 40 About 20 other extensions
> 
> Can anyone confirm they see this too?
> 
> Not sure if this only happens on Gnome 46 or on older versions as well.

Could you please file a seperate issue, thank you ❤️

**Comment by iconoclasthero** [comment]
I'm not sure this is the same error:
```
$ sudo apt install gir1.2-gda-5.0 gir1.2-gsound-1.0
gir1.2-gda-5.0 is already the newest version (5.2.10-4.1build1).
gir1.2-gsound-1.0 is already the newest version (1.0.3-3.2build2).
```
```
$ neofetch
OS: Rhino Linux 2024.1 (server) x86_64 
Kernel: Linux 6.9.1-060901-generic 
DE: GNOME 46.2 (Wayland) 
WM: Mutter 
```

![image](https://github.com/oae/gnome-shell-pano/assets/535962/5ec6d781-85d5-4c4e-ba73-6908803f4dbd)

**Comment by Seqaeon** [comment]
![Screenshot from 2024-07-24 19-46-42](https://github.com/user-attachments/assets/0db39360-2748-4878-8aae-6953300d8efe)

![image](https://github.com/user-attachments/assets/c016eaa3-cb63-4dad-a403-38d1cc17e1a4)


Stilll not working in GNOME 46

**Comment by teohhanhui** [comment]
@Seqaeon You have to use https://github.com/oae/gnome-shell-pano/releases/tag/v23-alpha2

**Comment by Seqaeon** [comment]
That worked. Thanks a lot @teohhanhui. 


A small question, it seems to copy everything I double click and I can't seem to see an option to get it to only copy stuff I press CTRL +C on

**Comment by teohhanhui** [comment]
> it seems to copy everything I double click and I can't seem to see an option to get it to only copy stuff I press CTRL +C on

It's in the settings. You'd want to use an extension manager like https://flathub.org/apps/com.mattjakeman.ExtensionManager (the best) to be able to access extension settings in general. But for Pano, you can right click on the widget and go to Settings.

**Comment by Seqaeon** [comment]
> > it seems to copy everything I double click and I can't seem to see an option to get it to only copy stuff I press CTRL +C on
> 
> It's in the settings. You'd want to use an extension manager like https://flathub.org/apps/com.mattjakeman.ExtensionManager (the best) to be able to access extension settings in general. But for Pano, you can right click on the widget and go to Settings.

I was using the gnome-extension app to edit its settings but couldn't figure out how to make it stop.
![image_2024-07-25_00-54-32](https://github.com/user-attachments/assets/274e13be-8b2e-4d15-b9da-f63bd51b6769)
![image_2024-07-25_00-54-21](https://github.com/user-attachments/assets/b80a0d22-7975-4cee-bb46-9d905e894260)

**Comment by hockdudu** [comment]
It's under "sync primary"

**Comment by Seqaeon** [comment]
> It's under "sync primary"

I disabled that and it didn't change

**Comment by Seqaeon** [comment]
> It's under "sync primary"


![image_2024-07-25_10-04-05](https://github.com/user-attachments/assets/06f973ef-262b-4516-bcb6-c359afc80b30)


See trying to select some stuff and it starts copying straight from the first character selected.

**Comment by hockdudu** [comment]
Weird... This setting should disable exactly that. Are the settings being actually persisted? What's the output of the following command?

```shell
dconf dump /org/gnome/shell/extensions/pano/
```

It should output this line, among others

```shell
sync-primary=false
```

**Comment by Seqaeon** [comment]
> Weird... This setting should disable exactly that. Are the settings being actually persisted? What's the output of the following command?
> 
> ```shell
> dconf dump /org/gnome/shell/extensions/pano/
> ```
> 
> It should output this line, among others
> 
> ```shell
> sync-primary=false
> ```


`[/]
global-shortcut=['<Super>v']
history-length=500
icon-pack=uint32 0
incognito-shortcut=['<Control><Alt><Super>v']
item-size=200
paste-on-select=true
sync-primary=false
window-position=uint32 0`

**Comment by Seqaeon** [comment]
> Strange... Does this persist when you restart the computer? Maybe is another extension or tool conflicting with this extension, could you try disabling other extensions and seeing if this persists?
> 
> If this doesn't work, I believe it's best to open a new issue, as this seems to be a bug.

Wow, it's been fixed, the issue seems to be I had GPaste also installed.

**Comment by zvictor** [comment]
For anyone interested, this is how I managed to run [v23-alpha2](https://github.com/oae/gnome-shell-pano/releases/tag/v23-alpha2) in Nix:

* `gnome-shell-extension-pano.patch`:
```diff
--- a/extension.js
+++ b/extension.js
@@ -1,3 +1,7 @@
+import GIRepository from 'gi://GIRepository';
+GIRepository.Repository.prepend_search_path('@gda_path@');
+GIRepository.Repository.prepend_search_path('@gsound_path@');
+
 import Gio from 'gi://Gio';
 import GLib from 'gi://GLib';
 import * as extension_js from 'resource:///org/gnome/shell/extensions/extension.js';
```

* `gnome-shell-extension-pano.nix`:
```nix
{ stdenv, fetchzip, lib, gnome, glib, libgda, gsound, substituteAll, wrapGAppsHook }:

stdenv.mkDerivation rec {
  pname = "gnome-shell-extension-pano";
  version = "v23-alpha2";
  
  src = fetchzip {
    url = "https://github.com/oae/gnome-shell-pano/releases/download/${version}/pano@elhan.io.zip";
    sha256 = "Y8WgVUHX094RUwYKdt7OROPZMl3dakK0zOU9OTdyqxc=";
    stripRoot = false;
  };

  patches = [
    (substituteAll {
      src = ./gnome-shell-extension-pano.patch;
      gsound_path = "${gsound}/lib/girepository-1.0";
      gda_path = "${libgda}/lib/girepository-1.0";
    })
  ];

  buildInputs = [
    gnome.gnome-shell
    libgda
    gsound
  ];

  nativeBuildInputs = [
    wrapGAppsHook
  ];

  installPhase = ''
    runHook preInstall
    local_ext_dir=$out/share/gnome-shell/extensions/pano@elhan.io
    install -d $local_ext_dir
    cp -r * $local_ext_dir

    # Ensure typelibs are directly accessible
    mkdir -p $out/lib/girepository-1.0
    ln -s ${gsound}/lib/girepository-1.0/* $out/lib/girepository-1.0/
    ln -s ${libgda}/lib/girepository-1.0/* $out/lib/girepository-1.0/

    runHook postInstall
  '';

  meta = with lib; {
    description = "Pano GNOME Shell Clipboard Management Extension (${version} pre-release)";
    homepage = "https://github.com/oae/gnome-shell-pano";
    license = licenses.gpl2Plus;
    platforms = platforms.linux;
    maintainers = [ maintainers.zvictor ];
  };
```

**Comment by EmreKaratopuk** [comment]
Thank you for the pre-release, it works very well. Is there an ETA for full Gnome 46 support?

**Comment by carmen-gh** [comment]
and here comes Gnome 47 with latest Fedora release... I hope I can use it someday again :)

**Comment by Tanguille** [comment]
> and here comes Gnome 47 with latest Fedora release... I hope I can use it someday again :)

I am using the pre-release on fedora 41 without issues so far.

**Comment by Malix-Labs** [comment]
If it is working without (or with minor) issues, could you _please_ [push it to EGO](https://extensions.gnome.org/extension/5278/pano/) ?

**Comment by Totto16** [comment]
> If it is working without (or with minor) issues, could you _please_ [push it to EGO](https://extensions.gnome.org/extension/5278/pano/) ?

That's not possible, as we have some external dependencies, that they no longer accept, for more information see  https://github.com/oae/gnome-shell-pano/issues/271#issuecomment-2093444107

**Comment by teohhanhui** [comment]
> Correct, the "issue" with highlight.js is, that EGO doesn't want to have that too much third party dependencies in published extensions.

They don't accept bundled dependencies? :scream:

**Comment by Malix-Labs** [comment]
See https://github.com/oae/gnome-shell-pano/pull/278#issuecomment-2453482576

TL;DR: apparently, the blocker to make Pano available in [GNOME Shell Extensions](https://extensions.gnome.org/) again is too complex and only @Totto16 and @oae could solve it and make it mainstream again

**Comment by Totto16** [comment]
> > Correct, the "issue" with highlight.js is, that EGO doesn't want to have that too much third party dependencies in published extensions.
> 
> They don't accept bundled dependencies? 😱

Some small ones, but not something like [highlightjs](https://highlightjs.org/). 😓 

We use that package for code highlighting 😓 

Since I started #278 a half year ago, we also could just remove that functionality for now, and then re-add that later, when I managed to do that. For context, we can use subprocesses, that call to an executable, that a user can optionally install, but doing that correctly and without huge cpu load (when we scan up to 500 clipboard contents, to determine if they are code and then highlight them if they are) is not that easy 😓 

I'll look into this this week, since deactivating functionailuty isn't that complicated.

**Comment by LeonardoMantovani** [comment]
I know this is kind of an hard thing to fix, but any news about the workaround described by @Totto16 in the previous comment?
Idk about other users, but personally even though I used to enjoy the code highlighting feature, I can definitely live without that until the complex things get sorted out if this means having back Pano in the short term

**Comment by venteto** [comment]
> Idk about other users, but personally even though I used to enjoy the code highlighting feature, I can definitely live without that until the complex things get sorted out if this means having back Pano in the short term

I agree, anything beyond basic clipboard management functionality that is blocking an update to Pano should be removed for now, and also perhaps just added to a separate "Advanced" version of Pano that people could opt to install rather than "Basic" Pano. Code highlighting is of zero use to me.

**Comment by oae** [comment]
At this point, we can drop the code highlighting, merge the UI overhaul #280 , and publish it.

**Comment by pavitra-infocusp** [comment]
Waiting for Ubuntu 24.04 support!

**Comment by andrejskuidins** [comment]
> Waiting for Ubuntu 24.04 support!

I would better say Ubuntu 25.04!! ;)

**Comment by emilfihlman** [comment]
What's the status and holdup? Isn't everything ready for 48 with the the changes now?

**Comment by Georgiy-JO** [comment]
Sorry for bothering, could not find the way to fix the issue myself.
System: Ubuntu 24.04.3 LTS
GNOME 46
Issue: Pano stopped working. 
About a week ago Pano stopped reviving any copied text or files. Basically it does not matter what I copy, system copying works fine, but nothing is added to the Pano. For a while it just been keeping old instances, but after me cleaning the list, it has just been empty.
I tried switching to Xorg and reinstalling Pano itself (both v22   and several alphas of v23) nothing changed.
Interestingly, Clipboard Indicator works fine, but I would like to find a way to fix Pano (it's lovely).
Some AI tools recommended installing/reinstalling xclip, xsel, wl-clipboard, but naturally it didn't help. 

I wonder if the issue is well-known and if there are another ways to fix it that I could try.
I can provide some extra information if it would be required.

Thank you in advance for your help.

**Comment by talesam** [comment]
> Sorry for bothering, could not find the way to fix the issue myself. System: Ubuntu 24.04.3 LTS GNOME 46 Issue: Pano stopped working. About a week ago Pano stopped reviving any copied text or files. Basically it does not matter what I copy, system copying works fine, but nothing is added to the Pano. For a while it just been keeping old instances, but after me cleaning the list, it has just been empty. I tried switching to Xorg and reinstalling Pano itself (both v22 and several alphas of v23) nothing changed. Interestingly, Clipboard Indicator works fine, but I would like to find a way to fix Pano (it's lovely). Some AI tools recommended installing/reinstalling xclip, xsel, wl-clipboard, but naturally it didn't help.
> 
> I wonder if the issue is well-known and if there are another ways to fix it that I could try. I can provide some extra information if it would be required.
> 
> Thank you in advance for your help.

I generate package for my distro from gitgub, if you want you can try to convert the package to deb and install it.
https://github.com/big-comm/gnome-shell-extension-pano

**Comment by rostmodern** [comment]
Do the main functions work reliably on Ubuntu 24.04 with Wayland / Gnome 46 these days? 
Any major drawbacks of the V23 alpha5 version? 

would love to hear your feedback! 
p.s. code highlighting is not important to me
